etcdserver/api/membership/cluster.go

```diff
@@ -759,16 +759,21 @@ func ValidateClusterAndAssignIDs(lg *zap.Logger, local *RaftCluster, existing *R
 	if len(ems) != len(lms) {
 		return fmt.Errorf("member count is unequal")
 	}
-	sort.Sort(MembersByPeerURLs(ems))
-	sort.Sort(MembersByPeerURLs(lms))
 
 	ctx, cancel := context.WithTimeout(context.TODO(), 30*time.Second)
 	defer cancel()
 	for i := range ems {
-		if ok, err := netutil.URLStringsEqual(ctx, lg, ems[i].PeerURLs, lms[i].PeerURLs); !ok {
+		var err error
+		ok := false
+		for j := range lms {
+			if ok, err = netutil.URLStringsEqual(ctx, lg, ems[i].PeerURLs, lms[j].PeerURLs); ok {
```

Comment:
The method `netutil.URLStringsEqual` and its callees `netutil.urlEqual` and `netutil.resolveTCPAddrs` take a pair of `[]url.URL` , so perhaps a single loop is assumed. By dividing into multiple level loop, same URL is potentially getting resolved multiple times.
I am not familiar with a real setup to see if it leads to inefficiency and we need to cache the result.

Reply:
See above comment, should already be cached by the lower layers (host/resolver). Also see this request to implement a cache in golang that was declined for the same reason https://github.com/golang/go/issues/24796#issuecomment-383716244